port/unix/omrmmap.c

```diff
@@ -50,11 +49,93 @@
 #include "protect_helpers.h"
 #include "omrportpriv.h"
 
+#if defined(J9ZOS390)
+#include "omrvmem.h"
+#if !defined(MAP_FAILED)
+#define MAP_FAILED ((void *)(intptr_t)-1)
+#endif /* !defined(MAP_FAILED) */
+#endif /* defined(J9ZOS390) */
+
 #if defined(AIXPPC)
 #include <sys/shm.h>
 #include <sys/vminfo.h>
 #endif/*AIXPPC*/
 
+#if defined(J9ZOS390)
+static J9MmapHandle *
+omrmmap_read_map_file(struct OMRPortLibrary *portLibrary, intptr_t file, uint64_t offset, uintptr_t size, uint32_t category)
+{
+	/* default implementation will allocate memory and read the file into it */
+	uintptr_t numBytesRead = 0;
+	intptr_t rc = 0;
+	void *mappedMemory = NULL;
+	void *allocPointer = NULL;
+	J9MmapHandle *returnVal = NULL;
+
+	Trc_PRT_mmap_map_file_default_entered(file, size);
+
+	if (-1 == file) {
+		Trc_PRT_mmap_map_file_default_badfile();
+		return NULL;
+	}
+
+	if (0 == size) {
+		struct stat buf;
+
+		memset(&buf, 0, sizeof(buf));
+		if (-1 == fstat(file - FD_BIAS, &buf)) {
+			Trc_PRT_mmap_map_file_unix_filestatfailed();
+			portLibrary->error_set_last_error(portLibrary, errno, OMRPORT_ERROR_MMAP_MAP_FILE_STATFAILED);
+			return NULL;
+		}
+		size = buf.st_size;
+	}
+
+	if (0 != portLibrary->file_seek(portLibrary, file, offset, EsSeekSet)) {
+		Trc_PRT_mmap_map_seek_failed(offset);
+		return NULL;
+	}
+
+	allocPointer = portLibrary->mem_allocate_memory(portLibrary, size, OMR_GET_CALLSITE(), category);
+	Trc_PRT_mmap_map_file_default_allocPointer(allocPointer, size);
+
+	if (NULL == allocPointer) {
+		Trc_PRT_mmap_map_file_default_badallocPointer();
+		portLibrary->file_close(portLibrary, file);
+		return NULL;
+	}
+
+	mappedMemory = allocPointer;
+	Trc_PRT_mmap_map_file_default_mappedMemory(mappedMemory);
+
+	numBytesRead = 0;
+	rc = 0;
+	while (size > numBytesRead) {
+		rc = portLibrary->file_read(portLibrary, file, (void *)(((uintptr_t)mappedMemory) + numBytesRead), size - numBytesRead);
+		if (0 >= rc) {
+			/* failed to completely read the file */
+			Trc_PRT_mmap_map_file_default_badread();
+			portLibrary->mem_free_memory(portLibrary, allocPointer);
+			return NULL;
+		}
+		numBytesRead += rc;
+		Trc_PRT_mmap_map_file_default_readingFile(numBytesRead);
+	}
+
+	returnVal = (J9MmapHandle *)portLibrary->mem_allocate_memory(portLibrary, sizeof(J9MmapHandle), OMR_GET_CALLSITE(), category);
+	if (NULL == returnVal) {
+		Trc_PRT_mmap_map_file_cannotallocatehandle();
```

Comment:
Same here: an "exit" tracepoint is needed.

Reply:
Updated to use` TraceExit-Exception` trace point here.